File: java/src/Bj2116.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Bj2116 {
    static void solve() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        List<Integer>[] dices = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            String[] diceLine = br.readLine().split(" ");
            dices[i] = new ArrayList<>();
            for (String value : diceLine) {
                dices[i].add(Integer.parseInt(value));
            }
        }

        int ans = 0;

        for (int i = 0; i < 6; i++) {
            int downIndex = i;
            int max = 0;
            int downValue = dices[0].get(i);
            for (List<Integer> dice : dices) {
                downIndex = dice.indexOf(downValue);
                int upIndex = getOpponent(downIndex);
                int finalDownValue = downValue;
                max += dice.stream()
                    .filter((value) -> value != finalDownValue && dice.indexOf(value) != upIndex)
                    .max(Integer::compare).get();
                downValue = dice.get(upIndex);
            }
            ans = Math.max(ans, max);
        }

        System.out.println(ans);
    }

    private static int getOpponent(int up) {
        switch (up) {
            case 0:
                return 5;
            case 1:
                return 3;
            case 2:
                return 4;
            case 3:
                return 1;
            case 4:
                return 2;
            case 5:
                return 0;
            default:
                return 0;
        }
    }
}

File: kotlin/src/Bj2116.kt
import java.io.BufferedReader
import java.io.InputStreamReader

class Bj2116 {
    companion object {
        fun solve() {
            val br = BufferedReader(InputStreamReader(System.`in`))
            val n = br.readLine().toInt()
            val dices = Array(n) { listOf<Int>() }
            for (i in 0 until n) {
                dices[i] = br.readLine().split(" ").map { it.toInt() }
            }

            var ans = 0

            dices.first().mapIndexed{ index, it ->
                var downIndex: Int = index
                var max = 0
                var downValue = it
                dices.map { dice ->
                    downIndex = dice.indexOf(downValue)
                    val upIndex = getOpponent(downIndex)
                    max += dice.filterIndexed { index, i ->
                        index != downIndex && index != upIndex
                    }.max()
                    downValue = dice[upIndex]
                }
                ans = if (ans < max) max else ans
            }
            println(ans)
        }

        private fun getOpponent(up: Int) = when (up) {
            0 -> 5
            1 -> 3
            2 -> 4
            3 -> 1
            4 -> 2
            5 -> 0
            else -> {
                0
            }
        }
    }
}